mani_skill/envs/tasks/tabletop/pull_cube_tool.py

```diff
@@ -0,0 +1,278 @@
+            color=np.array([12, 42, 160, 255]) / 255,
+            name="cube",
+            body_type="dynamic",
+        )
+
+        # Create and position the L-shaped tool in the scene
+
+        self.l_shape_tool = self._build_l_shaped_tool(
+            handle_length=self.handle_length,
+            hook_length=self.hook_length,
+            width=self.width,
+            height=self.height,
+        )
+        self.l_shape_tool.set_pose(
+            sapien.Pose(p=[-0.1, -0.1, self.cube_half_size + self.height])
+        )
+
+    def _initialize_episode(self, env_idx: torch.Tensor, options: dict):
+        with torch.device(self.device):
+            b = len(env_idx)
+            self.scene_builder.initialize(env_idx)
+
+            # Initialize robot position and orientation
+            qpos = torch.tensor(
+                [
+                    0.0,
+                    np.pi / 8,
+                    0,
+                    -np.pi * 5 / 8,
+                    0,
+                    np.pi * 3 / 4,
+                    np.pi / 4,
+                    0.04,
+                    0.04,
+                ]
+            )
+
+            qpos = (
+                torch.normal(
+                    0, self.robot_init_qpos_noise, (b, len(qpos)), device=self.device
+                )
+                + qpos
+            )
+
+            qpos[:, -2:] = 0.04
+            self.agent.robot.set_qpos(qpos)
+            self.agent.robot.set_pose(sapien.Pose([-0.615, 0, 0]))
+
+            # Initialize the tool
+
+            tool_xyz = torch.zeros((b, 3))
+            tool_xyz[..., :2] = (
+                torch.rand((b, 2)) * 0.2 - 0.1
+            )  # spawn tool in region where x,y in [-0.1, 0.1]
+            tool_xyz[..., 2] = self.height / 2  # place tool on table
+            tool_q = [1, 0, 0, 0]  # no rotation
+
+            tool_pose = Pose.create_from_pq(p=tool_xyz, q=tool_q)
+            self.l_shape_tool.set_pose(tool_pose)
+
+            # Initialize the cube a bit away from the base of the arm
+
+            cube_xyz = torch.zeros((b, 3))
+            cube_xyz[..., 0] = self.arm_reach + torch.rand(b) * (
+                self.handle_length - 0.08
+            )  # Just outside arm's reach
+            cube_xyz[..., 1] = torch.rand(b) * 0.4 - 0.2  # Random y position
+            cube_xyz[..., 2] = self.cube_size / 2  # Place on the table
+
+            cube_q = randomization.random_quaternions(
+                b,
+                lock_x=True,
+                lock_y=True,
+                lock_z=False,
+                bounds=(-np.pi / 6, np.pi / 6),
+            )
+
+            cube_pose = Pose.create_from_pq(p=cube_xyz, q=cube_q)
+            self.cube.set_pose(cube_pose)
+
+    def _get_obs_extra(self, info: Dict):
+        obs = dict(
+            tcp_pose=self.agent.tcp.pose.raw_pose,
+            cube_pose=self.cube.pose.raw_pose,
+            tool_pose=self.l_shape_tool.pose.raw_pose,
+        )
```

Comment:
Following the setup of other envs (e.g. push cube), can you ensure the ground truth info (tool pose, cube pose) are only provided if the observation is state based

Reply:
this is still not resolved. cube pose and tool pose are always included